src/org/rascalmpl/dap/SuspendedState.java

```diff
@@ -142,4 +146,84 @@ public void addVariable(RascalVariable variable){
         variables.put(nextReferenceID, variable);
     }
 
+    public RascalVariable setVariable(int referenceID, String name, String valueStr) {
+        if(!scopes.containsKey(referenceID)){
+            return null;
+        }
+        IRascalFrame frame = scopes.get(referenceID);
+
+        // First: check that this variable exists in the suspended frame. If it
+        // doesn't, we don't create it silently; signal failure so the UI can
+        // report the error and avoid temporarily showing a value that will be
+        // rolled back later.
+        if (!frame.getFrameVariables().contains(name)) {
+            return null;
+        }
+
+        // Evaluate the new value in a fork to avoid touching the suspended evaluator's
+        // execution state. Then wrap the raw IValue into a Result bound to the
+        // suspended evaluator so it can be stored in its Environment.
+        Result<IValue> forkResult = evaluator.fork().eval(evaluator.getMonitor(), valueStr, URIUtil.rootLocation("internal"));
```

Comment:
@jurgenvinju what are the guarantees about fork and it's side-effect in other modules? I always assumed that `fork` is full of subtle issues?

Reply:
yes, fork only works under very well controlled conditions which we do not have here. One of them is that the program's modules that are forked do not contain any mutable global variables, if my memory is not fooling me.

I talked to @jbdoderlein about this part. We can write a different version of `eval` which executes with a given `Environment` as scope parent, and return the state of the evaluator as before after we have finished. All we have to assume is that the current interpreter is suspended as it is evaluating the breakpoint or step interaction.